### PR TITLE
Issue #4193: The `file_temporary_path` is not set on `system.core.json` file

### LINKED ISSUE #4193: Set `file_temporary_path` in initial `system.core.json` file
**Description of the bug**

Despite the `/admin/config/media/file-system` page on a freshly installed Backdrop site shows that **Temporary directory** is set to `/tmp`:

Screen Shot 2019-11-03 at 3 46 43 AM

the `file_temporary_path` variable in the `system.core.json` file in active configuration directory shows empty. Therefore the function `config_get('system.core', 'file_temporary_path')` returns empty value causing number of serious troubles.

**Steps To Reproduce**

Just deploy a new Backdrop site and see the described difference between UI and configuration file.

You can quickly fire up the `cat files/config_*/active/system.core.json | grep file_temporary_path` command on CLI to see:

```
    "file_temporary_path": "",
```

At the same time, if you go to the `/admin/config/media/file-system` page then it shows `/tmp` for Temporary directory.

**Expected behavior**

The `file_temporary_path` variable in `system.core.json` file should match what the website's user interface shows for **Temporary directory**.

**Additional information**

If you save the `/admin/config/media/file-system` page, then the `file_temporary_path` variable in `system.core.json` file gets correct value. However, we should not expect that every user necessarily goes to and saves the `/admin/config/media/file-system` page just after installation. 

To me this seems pretty urgent issue to fix, so I'll be submitting PR soon.

---
~~PR by @alanmels: https://github.com/backdrop/backdrop/pull/2971~~
~~New PR by @alanmels: https://github.com/backdrop/backdrop/pull/2972~~
Yet another PR by @alanmels: https://github.com/backdrop/backdrop/pull/2988

### PR DESCRIPTION
**Pull request by alanmels**
Fixes https://github.com/backdrop/backdrop-issues/issues/4193

### CODE REVIEW

**Review by herbdool — COMMENTED** [review]
This looks good to me.

File: core/includes/file.inc [hunk 1]
```diff
@@ -2675,6 +2675,10 @@ function file_directory_temp() {
       // everything to use slash which is supported on all platforms.
       $temporary_directory = str_replace('\\', '/', $temporary_directory);
     }
+    
+    // Now when the temporary directory is found, we need to save it to
+    // the configuration file.
+    config_set('system.core', 'file_temporary_path', $temporary_directory);
```
Comment (indigoxela) [inline core/includes/file.inc:2681]: Setting the tmp dir shouldn't happen in function file_directory_temp() in file.inc, as that function is called several times, which means, the config item gets set again and again.

See: https://api.backdropcms.org/api/backdrop/core%21includes%21file.inc/function/file_directory_temp/1

It should rather go to core/modules/system/system.install to make sure, it only happens once on install.

Comment (alanmels) [reply core/includes/file.inc:2681]: That's exactly why @jenlampton has put this comment https://github.com/backdrop/backdrop-issues/issues/4193#issuecomment-550064054 So we should call it only once - during the install - and all the other instances should replace with `config_get()`.

However, even if we didn't try to replace other calls to the function, I don't agree that it  

> gets set again and again

because there is a check in the very beginning of the module:

```
function file_directory_temp() {
  $temporary_directory = config_get('system.core', 'file_temporary_path');
  if (empty($temporary_directory)) {
```

In other words, the execution never proceeds if temp directory has already been put into the config file. So nothing terrible in case of multiple calls too.

Comment (indigoxela) [reply core/includes/file.inc:2681]: > So nothing terrible in case of multiple calls too.

True, but there's no need to do it here. Running config_set once at install would be much cleaner. My recommendation is still to _not_ do this in file.inc but in the installer.

Comment (alanmels) [reply core/includes/file.inc:2681]: Well, I'll be more than happy if core gets more optimized. The multiple checks have already been there. They are not caused by my PR. 

My PR is trying to deal with specific problem that Backdrop UI is giving wrong information to user. But let's see how this goes - if core maintainers would be willing to introduce multiple changes instead one suggested.

Your idea to do checks on installer is quite logical. In fact, if the function gets called only once by installer, then there is no need for it in the module - the whole function needs to move to installer.

Comment (indigoxela) [reply core/includes/file.inc:2681]: > the whole function needs to move to installer.

Function file_directory_temp() is a utility and at its perfect place.
Please read at least the first few sentences of the api description of function file_directory_temp().

Doing the _initial config_set_ of the tmp dir should go to the installer.

Comment (alanmels) [reply core/includes/file.inc:2681]: > > the whole function needs to move to installer.
> 
> Function file_directory_temp() is a utility and at its perfect place.
> Please read at least the first few sentences of the api description of function file_directory_temp().
> 
> Doing the _initial config_set_ of the tmp dir should go to the installer.

@indigoxela The above discussion is little outdated, because after you explained your idea on gitter I had created second PR just like you wanted.

Comment (herbdool) [reply core/includes/file.inc:2681]: The reason I thought this PR was okay was because, it makes a pretty small change and I checked D7 version and it does the same thing: saves the constructed variable if it's not already saved. It should only happen once no matter how many times the function is called. (Putting it in system.install may be cleaner but perhaps this PR helps in case where someone manually deletes the variable so it can automatically recreate it.)

And that's the only change I think is needed: I don't think we should try to replace the use of this function. (This old d.org issue https://www.drupal.org/node/930122 describes that they intentionally used this function instead of calling the variable directly.)

Comment (alanmels) [reply core/includes/file.inc:2681]: > Putting it in system.install may be cleaner but perhaps this PR helps in case where someone manually deletes the variable so it can automatically recreate it.

Interesting catch. There will be some folks who get dirty their hands deleting entries in the config file. Then we are back to the situation when UI shows a path that is not present in the config file. So my first PR might still be better than the second one.

**Comment by alanmels** [comment]
Closed per https://github.com/backdrop/backdrop-issues/issues/4193#issuecomment-552068391
